### DISCUSSION TITLE
4D noise

### DISCUSSION THREAD

**Discussion by unicomp21** [stegu/psrdnoise#9] [Q&A] [answered]
@stegu if I need 4D noise, what should I use?

**Comment by stegu** [comment]
I have actually stopped using (and implementing) 4D noise after I made 3D
"flow noise", which animates without a translation, is cheap to compute and
also tiles nicely. 4D simplex noise using the naive extension of the 3D
simplex grid is a bit too sparse on the vertices and looks washed out with
large swaths being close to zero, and classic 4D noise looks very different
depending on where and how you sample it. There is probably a more uniform
simplex grid in 4D to try, but I haven't looked further into that. All
other grids have a less straightforward inverse mapping. The 3D OpenSimplex
noise is a good example of that.

The look of 4D simplex noise can be improved by simply adding a few
carefully displaced terms together, or taking smaller steps than one full
octave in fractal sums to add more wiggles.

If you can state your requirements a little more in detail, perhaps by
explaining your application in general terms, I might be able suggest
something more specific. If there is a strong demand for 4D noise, I should
look into making a modern version.



Den mån 19 sep. 2022 13:34John Davis ***@***.***> skrev:

**Reply by unicomp21** [reply]
In a nutshell, I'm simply wanting to animate 3D noise across time.  Maybe I can do the same w/ flow noise?

**Comment by stegu** [comment, marked as answer]
Yes, definitely. Animate the "phase" (gradient rotation) parameter in our
3D psrdnoise, and you have a visual effect similar to a translation of a 3D
slice of 4D noise. The animation is periodic, which may or may not be what
you want, but the periodicity can be easily masked in a fractal sum by
using irrational relations between the rotation speeds of different terms.
You can sneak in a slow 3D translation as well -- it will be masked by the
rapid motion from the gradient rotations.

Den mån 19 sep. 2022 17:06John Davis ***@***.***> skrev:

**Comment by unicomp21** [comment]
Thanks @stegu !

**Comment by ezequielmezam** [comment]
Hello Stefan,

In psrdnoise3.wgsl the struct NG3 declaration, the semicolon can be changed to comma:

struct NG3 {
   noise: f32,
  gradient: vec3<f32>
};   //optional semicolon

**Comment by stegu** [comment]
"Can" or "should"? My knowledge of WGSL is only based on examples and what
works on the platforms I tested it on. If the correct formal syntax is a
comma and no semicolon at the end, I'll change it, of course. If the formal
syntax allows both, and this isn't just a compiler anomaly that I'm
abusing, I'll leave it.

Please advise. I'm confused by all these different shading languages having
different bur similar syntaxes.

  /Stefan G
